src/attributes.rs

```diff
@@ -0,0 +1,282 @@
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it starts with neither `#[` nor `#![`."
+                ),
+                raw_trimmed
+            )
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AttributeMetaItem<'a> {
+    pub raw_item: &'a str,
+    pub base: &'a str,
+    pub assigned_item: Option<&'a str>,
+    pub arguments: Option<Vec<Rc<AttributeMetaItem<'a>>>>,
+}
+
+impl<'a> AttributeMetaItem<'a> {
+    /// Tries to parse `raw` as a comma-separated sequence of `AttributeMetaItem`'s
+    /// wrapped in parentheses, square brackets or curly brackets.
+    fn slice_arguments(raw: &'a str) -> Option<Vec<Rc<AttributeMetaItem<'a>>>> {
+        let is_left = |c| c == '(' || c == '[' || c == '{';
+        let is_right = |c| c == ')' || c == ']' || c == '}';
+        let matching_right = |c| match c {
+            '(' => ')',
+            '[' => ']',
+            '{' => '}',
+            _ => unreachable!("Tried to find matching right bracket for {}.", c),
+        };
+
+        let raw_trimmed = raw.trim();
+        let first_char = raw_trimmed.chars().next()?;
+        let raw_meta_seq = raw_trimmed
+            .strip_prefix(is_left)?
+            .strip_suffix(|c| c == matching_right(first_char))?
+            .trim();
+
+        let mut i = 0; // index of the first character after the last comma
+        let mut bracket_depth = 0; // number of currently opened brackets
+        let mut arguments: Vec<Rc<AttributeMetaItem>> = Vec::new(); // meta items constructed so far
+
+        for (j, c) in raw_meta_seq.chars().enumerate() {
+            if is_left(c) {
+                bracket_depth += 1;
+            } else if is_right(c) {
+                bracket_depth -= 1;
+            } else if c == ',' {
+                // We only do a recursive call when the comma is on the outermost level.
+                if bracket_depth == 0 {
+                    arguments.push(Rc::new(AttributeMetaItem::new(&raw_meta_seq[i..j])));
+                    i = j + 1;
+                }
+            }
+        }
+
+        // If the last comma was not a trailing one, there is still one meta item left.
+        if i < raw_meta_seq.len() {
+            arguments.push(Rc::new(AttributeMetaItem::new(&raw_meta_seq[i..])));
+        }
+
+        Some(arguments)
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let simple_path_char = |c: char| c.is_alphanumeric() || c == '_' || c == ':';
+        let raw_trimmed = raw.trim();
+
+        if let Some(path_end) = raw_trimmed.find(|c| !simple_path_char(c)) {
+            let simple_path = &raw_trimmed[0..path_end];
+            let attr_input = &raw_trimmed[path_end..];
```

Comment:
Can you double-check whether this is all in char-index coordinates or in byte coordinates? Not all chars in UTF-8 are one byte, and I'm worried about mixing up units.

Perhaps also consider adding some test cases that use UTF-8 chars that are valid in identifiers but are not one byte long in UTF-8, to make sure this all works correctly.